src/os_db.ml

```diff
@@ -158,294 +165,330 @@ let preregister_table =
+	in
+        lwt userid = Lwt_Query.view_one dbh
+	  <:view< {x = currval $users_userid_seq$} >>
         in
-        let userid = userid#!x in
-        lwt () =
-          Lwt_Query.query dbh
-            <:insert< $emails_table$ :=
-                        { email = $string:email$;
-                          userid  = $int64:userid$;
-                          validated = emails_table?validated}
-            >>
-        in
-        lwt () = remove_preregister email in
-        Lwt.return userid
+	let userid = userid#!x in
+	lwt () = Lwt_Query.query dbh
+	  <:insert< $emails_table$ :=
+           { email = $string:email$;
+             userid  = $int64:userid$;
+             validated = emails_table?validated
+           } >>
+	in
+	lwt () = remove_preregister email in
+	Lwt.return userid
       )
 
   let update ?password ?avatar ~firstname ~lastname userid =
-    full_transaction_block (fun dbh ->
-      (match password, avatar with
-        | None, None ->
-          Lwt_Query.query dbh
-             <:update< d in $users_table$ :=
-                      { firstname = $string:firstname$;
-                        lastname = $string:lastname$ } |
-                       d.userid = $int64:userid$
-             >>
-        | None, Some avatar ->
-          let avatar = Some <:value< $string:avatar$ >> in
-          Lwt_Query.query dbh
-             <:update< d in $users_table$ :=
-                      { firstname = $string:firstname$;
-                        lastname = $string:lastname$;
-                        avatar = of_option $avatar$ } |
-                       d.userid = $int64:userid$
-             >>
-        | Some "", _ -> Lwt.fail (Failure "empty password")
-        | Some password, None ->
-          let password = (fst !pwd_crypt_ref) password in
-          let password = Some <:value< $string:password$ >> in
-          Lwt_Query.query dbh
-             <:update< d in $users_table$ :=
-                      { firstname = $string:firstname$;
-                        lastname = $string:lastname$;
-                        password = of_option $password$ } |
-                       d.userid = $int64:userid$
-             >>
-        | Some password, Some avatar ->
-          let password = (fst !pwd_crypt_ref) password in
-          let password = Some <:value< $string:password$ >> in
-          let avatar = Some <:value< $string:avatar$ >> in
-          Lwt_Query.query dbh
-             <:update< d in $users_table$ :=
-                      { firstname = $string:firstname$;
-                        lastname = $string:lastname$;
-                        avatar = of_option $avatar$;
-                        password = of_option $password$
-                       } |
-                       d.userid = $int64:userid$
-             >>
-      ))
+    if password = Some "" then Lwt.fail_with "empty password"
+    else
+      let password = password => (
+	(fun p _ -> as_sql_string @@ fst !pwd_crypt_ref p),
+	(fun _ -> password_of)
```

Comment:
Using `=>` requires as much code as simply pattern matching on password, so why not just use the built-in and known to all construct?